crypto/x509v3/v3_ncons.c

```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
 
 }
 
+#define CN2DNSID_NOMEM   (-2)
+#define CN2DNSID_BADNAME (-1)
+
+static int cn2dnsid(ASN1_STRING *cn, unsigned char **dnsid)
+{
+    int utf8_length;
+    int i;
+    unsigned char *utf8_value;
+    int isdnsname = 0;
+
+    if ((utf8_length = ASN1_STRING_to_UTF8(&utf8_value, cn)) < 0)
+        return CN2DNSID_NOMEM;
+
+    /*
+     * Some certificates have had names that include a *trailing* NUL byte.
+     * Remove these harmless NUL characters. They would otherwise yield false
+     * alarms with the following embedded NUL check.
+     */
+    while (utf8_length > 0 && utf8_value[utf8_length-1] == '\0')
+        --utf8_length;
+
+    /* Reject *embedded* NULs */
+    if ((size_t)utf8_length != strlen((char *)utf8_value))
+        return CN2DNSID_BADNAME;
+
+    /*
+     * XXX: loose DNS name syntax, also check names with '_'
+     * Check DNS name syntax, any '-' or '.' must be internal,
+     * and on either side of each '.' we can't have a '-' or '.'.
+     *
+     * If the name has just one label, we don't consider it a DNS name.  This
+     * means that "CN=sometld" cannot be precluded by DNS name constraints, but
+     * that is not a problem.
+     */
+    for (i = 0; i < utf8_length; ++i) {
+        unsigned char c = utf8_value[i];
+
+        if ((c >= 'a' && c <= 'z')
+            || (c >= 'A' && c <= 'Z')
+            || (c >= '0' && c <= '9')
+            || c == '_')
+            continue;
+
+        /* Dot and hyphen cannot be first or last. */
+        if (i > 0 && i < utf8_length - 1) {
+            if (c == '-')
+                continue;
+            /*
+             * Next to a dot the preceding and following characters must not be
+             * another dot or a hyphen.  Otherwise, record that the name is
+             * plausible, since it has two or more labels.
+             */
+            if (c == '.'
+                && utf8_value[i-1] != '.' && utf8_value[i+1] != '.'
```

Comment:
Isn't it sufficient just to check that the preceding character wasn't '.'? If there's one after it we'll catch it next time around the loop.

Reply:
You're right, fixing, but checking the "next" rather than previous to catch any problem sooner.